external/storm-hdfs/src/main/java/org/apache/storm/hdfs/trident/HdfsState.java

```diff
@@ -69,63 +78,92 @@
 
         abstract void doPrepare(Map conf, int partitionIndex, int numPartitions) throws IOException;
 
-        protected void rotateOutputFile() throws IOException {
+        abstract long getCurrentOffset() throws  IOException;
+
+        abstract void doCommit(Long txId) throws IOException;
+
+        abstract void doRecover(Path srcPath, long nBytes) throws Exception;
+
+        protected void rotateOutputFile(boolean doRotateAction) throws IOException {
             LOG.info("Rotating output file...");
             long start = System.currentTimeMillis();
-            synchronized (this.writeLock) {
-                closeOutputFile();
-                this.rotation++;
-
-                Path newFile = createOutputFile();
+            closeOutputFile();
+            this.rotation++;
+            Path newFile = createOutputFile();
+            if (doRotateAction) {
                 LOG.info("Performing {} file rotation actions.", this.rotationActions.size());
                 for (RotationAction action : this.rotationActions) {
                     action.execute(this.fs, this.currentFile);
                 }
-                this.currentFile = newFile;
             }
+            this.currentFile = newFile;
             long time = System.currentTimeMillis() - start;
             LOG.info("File rotation took {} ms.", time);
+        }
 
-
+        protected void rotateOutputFile() throws IOException {
+            rotateOutputFile(true);
         }
 
-        void prepare(Map conf, int partitionIndex, int numPartitions){
-            this.writeLock = new Object();
-            if (this.rotationPolicy == null) throw new IllegalStateException("RotationPolicy must be specified.");
+
+        void prepare(Map conf, int partitionIndex, int numPartitions) {
+            if (this.rotationPolicy == null) {
+                throw new IllegalStateException("RotationPolicy must be specified.");
+            } else if (this.rotationPolicy instanceof FileSizeRotationPolicy) {
+                long rotationBytes = ((FileSizeRotationPolicy) rotationPolicy).getMaxBytes();
+                LOG.warn("FileSizeRotationPolicy specified with {} bytes.", rotationBytes);
+                LOG.warn("Recovery will fail if data files cannot be copied within topology.message.timeout.secs.");
+                LOG.warn("Ensure that the data files does not grow too big with the FileSizeRotationPolicy.");
+            } else if (this.rotationPolicy instanceof TimedRotationPolicy) {
+                LOG.warn("TimedRotationPolicy specified with interval {} ms.", ((TimedRotationPolicy) rotationPolicy).getInterval());
+                LOG.warn("Recovery will fail if data files cannot be copied within topology.message.timeout.secs.");
+                LOG.warn("Ensure that the data files does not grow too big with the TimedRotationPolicy.");
+            }
```

Comment:
It might be nicer to move this code into the subclasses own methods: FileSizeRotationPolicy#prepare and TimedRotationPolicy#prepare.  These prepare methods can each call `super.prepare(conf, partitionIndex, numPartitions)`.